### PR TITLE
Refactor PgCopy to BatchPersister

### PR DESCRIPTION
**Pull request by steven-sheehy**
**Description**:
- Add an `@Upsertable` annotation for domain classes that use upsert logic
- Add a `CompositeBatchPersister` that abstracts away the mapping of domain classes to `BatchInserter` instances
- Change `BatchInserter` to be responsible for its own database connection
- Fix `AccountBalanceFileParserTest` not validating balance items
- Fix start date property and its cache not reset between test runs
- Move `hedera.mirror.importer.parser.{streamType}.bufferSize` to `hedera.mirror.importer.parser.bufferSize`
- Rename `PgCopy` to `BatchInserter`
- Rename `UpsertPgCopy` to `BatchUpserter`

**Related issue(s)**:

**Notes for reviewer**:
Passing the SQL `Connection` around classes is unnecessary since Spring already associates a connection with the current thread when a transaction is started. So as long as you're on the same thread as the transaction initiator you can call `DataSourceUtils.getConnection()` and always receive the same connection object.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorDateRangePropertiesProcessor.java [hunk 1]
```diff
@@ -207,6 +212,14 @@ public DateRangeFilter(Instant startDate, Instant endDate) {
             }
         }
 
+        public static DateRangeFilter empty() {
+            return new DateRangeFilter(Instant.EPOCH.plusNanos(1), Instant.EPOCH);
+        }
+
+        public static DateRangeFilter all() {
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorDateRangePropertiesProcessor.java:219]: nit: maxRange or something is more descriptive

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    public CompositeBatchPersister(DataSource dataSource,
+                                   MeterRegistry meterRegistry,
+                                   CommonParserProperties properties,
+                                   Collection<UpsertQueryGenerator> upsertQueryGenerators) {
+        AnnotatedTypeScanner annotatedTypeScanner = new AnnotatedTypeScanner(Entity.class);
+        Set<Class<?>> domainClasses = annotatedTypeScanner.findTypes(Upsertable.class.getPackageName());
+
+        for (Class<?> domainClass : domainClasses) {
+            Upsertable upsertable = AnnotationUtils.findAnnotation(domainClass, Upsertable.class);
+            BatchPersister batchPersister = null;
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:54]: nit: `BatchPersister` isn't used beyond the inserter so no need for the variable.
Just insert straight in the if/else
```suggestion
            if (upsertable != null) {
                 UpsertQueryGenerator generator = getUpsertQueryGenerator(upsertQueryGenerators, domainClass);
                 batchInserters.put(new BatchUpserter(domainClass, dataSource, meterRegistry, properties, generator));
             } else {
                 batchInserters.put(new BatchInserter(domainClass, dataSource, meterRegistry, properties));
             }
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public void persist(Collection<? extends Object> items) {
+        if (items == null || items.isEmpty()) {
+            return;
+        }
+
+        Object item = items.iterator().next();
+        BatchPersister batchPersister = batchInserters.get(item.getClass());
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:74]: Might wanna check for `null` item possibility.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersisterTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private ContractRepository contractRepository;
+
+    @Resource
+    private EntityRepository entityRepository;
+
+    @Test
+    @Transactional
+    void persist() {
+        Contract contract = domainBuilder.contract().get();
+        Entity entity = domainBuilder.entity().get();
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersisterTest.java:56]: nit: for completeness you should add 1 or 2 non upsertable domains.

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looks good, some suggestions

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorDateRangePropertiesProcessor.java (hunk 1 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorDateRangePropertiesProcessor.java:219]: It's an existing method that just got re-sorted via auto-format. Only seems to be used in a single test anyway.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java (hunk 2 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:54]: I had it like that originally but that makes the `put` not DRY but also causes the line to split across multiple making the code uglier, IMO.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersisterTest.java (hunk 4 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersisterTest.java]: Sure, swapped Entity with ContractResult

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java (hunk 3 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java]: Added

**Review by xin-hedera — APPROVED** [review]
LGTM

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java]: Gotcha

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java [hunk 5]
```diff
@@ -71,10 +71,11 @@ public void persist(Collection<? extends Object> items) {
         }
 
         Object item = items.iterator().next();
-        BatchPersister batchPersister = batchInserters.get(item.getClass());
+        Class<?> itemClass = item != null ? item.getClass() :null;
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:74]: Why not just exit early on null? Just like we do when items is null or empty?

**Review by Nana-EC — CHANGES_REQUESTED** [review]
LG, just think we should return early instead of blocking paring on a null item in the items list

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java (hunk 5 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:74]: I don't think we should silently ignore using a `return` statement like null or empty. Having a null in the item list is an unsupported scenario that indicates a coding error. Imagine if the it was `List.of(null, new Contract(), new Contract())` then the last two valid contracts would be discarded.

Unless you're saying we should have a separate if item is null then throw exception? I didn't want to duplicate the exception throwing code so went with the less code approach.

Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/CompositeBatchPersister.java:74]: Actually you're right. Safer this way too. Carry on

**Review by Nana-EC — APPROVED** [review]
LGTM

**Pull request merged**